packages/labs/src/multi-select/multiSelect.tsx

```diff
@@ -0,0 +1,299 @@
+                    onKeyDown={this.handleKeyDown}
+                    onKeyUp={this.state.isOpen && handleKeyUp}
+                >
+                    <TagInput
+                        inputProps={defaultInputProps}
+                        onRemove={this.handleTagRemove}
+                        {...tagInputProps}
+                        className={classNames("pt-multi-select", tagInputProps.className)}
+                        values={this.state.selectedItems.map(this.props.tagRenderer)}
+                    />
+                </div>
+                <div onKeyDown={handleKeyDown} onKeyUp={handleKeyUp}>
+                    <ul className={Classes.MENU} ref={listProps.itemsParentRef}>
+                        {this.renderItems(listProps)}
+                    </ul>
+                </div>
+            </Popover>
+        );
+    }
+
+    private renderItems({ activeItem, filteredItems, handleItemSelect }: IQueryListRendererProps<T>) {
+        const { itemRenderer, noResults } = this.props;
+
+        if (filteredItems.length === 0) {
+            return noResults;
+        }
+
+        return filteredItems.map((item, index) => itemRenderer({
+            index,
+            item,
+            handleClick: (e) => handleItemSelect(item, e),
+            isActive: item === activeItem,
+            isSelected: this.isItemSelected(item),
+        }));
+    }
+
+    private handleQueryChange = (event: React.FormEvent<HTMLInputElement>) => {
+        const query = event.currentTarget.value;
+        let nextState = { query };
+
+        if (query.length === 0 && this.props.openOnKeyDown) {
+            nextState = { ...nextState, isOpen: false };
+        }
+
+        this.setState(nextState);
+    }
+
+    private getSelectedItemIndex(item: T) {
+        return this.state.selectedItems.indexOf(item);
+    }
+
+    private isItemSelected(item: T) {
+        return this.getSelectedItemIndex(item) !== -1;
+    }
+
+    private selectItem(item: T) {
+        const { openOnKeyDown, resetOnSelect } = this.props;
+        let nextState = { selectedItems: [...this.state.selectedItems, item] };
+
+        if (openOnKeyDown) {
+            nextState = { ...nextState, isOpen: false };
+        }
+        if (resetOnSelect) {
+            nextState = { ...nextState, query: "" };
+        }
+
+        this.setState(nextState);
+    }
+
+    private deselectItem(index: number) {
+        this.setState({ selectedItems: this.state.selectedItems.filter((_, i) => i !== index) });
+    }
+
+    private handleTagRemove = (_item: string, index: number) => {
+        this.deselectItem(index);
+
+        Utils.safeInvoke(this.props.tagInputProps.onRemove, _item, index);
+    }
+
+    private handleItemSelect = (item: T, _event: React.SyntheticEvent<HTMLElement>) => {
+        // check whether the query string is in the item list
+        if (item !== undefined && this.state.activeItem !== undefined) {
+            if (!this.isItemSelected(item)) {
```

Comment:
how about getting the `selectedIndex` once and using it here and on line 233? then you don't need `isItemSelected` at all (it's a simple `>= 0` check anyway).